chapters/functions.tex

```diff
@@ -2392,6 +2392,9 @@ \subsection{Annotations for External Functions}\label{annotations-for-external-l
   \item
     A piece of C code directly defining \lstinline[language=C]!myfoo!.
     Since no other \lstinline!Include! annotation is expected to contain a definition of \lstinline[language=C]!myfoo!, it is not necessary to make the definition \lstinline[language=C]!static!.
+  \item
+    A small piece of C code defining \lstinline[language=C]!myfoo! as a wrapper around a function in a linked library that does the real job, where the wrapper forwards the arguments as well as passes function pointers for a selection of functions from \filename{ModelicaUtilities.h}.
+    This allows the linked library to be built and distributed without direct access to \filename{ModelicaUtilities.h} or any library providing the implementation of \filename{ModelicaUtilities.h}, and when the external function is used with a particular tool, it will use that tool's implementation of \filename{ModelicaUtilities.h}.
```

Comment:
Do we really need to include this case? Even when one wants to use functions from ModelicaUtilities.h another option is often to have the wrapper call the linked library and do error reporting in the wrapper based on the result. To me that is a lot easier to understand, and if we don't add it people might believe that they need to go do this complicated stuff - and if we include both it is getting fairly long-winded. (And using ModelicaError for error reporting can cause havoc if the called function allocates resources and don't realize that it ModelicaError will not return.)

Reply:
I added this case because I encountered it in multiple libraries, and for external objects with complex needs for error handling, I can really see why one wants to have external code that is compiled into a library, with only minimal amount of code included via the `Include` annotation.

Thus, while I can't say we really need this case, as it is just an example, I at least find it a much more relevant example than directly providing a function declaration, and a better coding style than using an `#include` directive to include a C source with function definitions.

Reply:
I can understand the need for something like that text.

The issues I have with it are:
- It is a forward reference for ModelicaUtilities.h; it would be better to move that text to that section - less confusion for people just reading about include - and more helpful for people looking up those functions.
- It is somewhat esoteric - so it needs more context

The reason it is esoteric is that the first use of functions from ModelicaUtilities.h is just to call them (writing "called my_foo" below). If you need to send in a function it is often the case that you do more e.g.:
```
static void logger(const char*txt) {
   ModelicaFormatMessage("Foo-logger: %s", txt);
}
static void my_foo(double*x) {
  ModelicaMessage("called my_foo");
  library_foo(x, logger);
}
```
(I believe that there have been some specific issues with calling conventions for ModelicaUtilities.h that implied that you could not send in ModelicaMessage directly as logger. I'm not sure if that is still an issue that we need to document.) Regardless if you want adapt the arguments or write more you anyway have to write the wrapper; it also creates a separation of interfaces (e.g., the logging-function in FMI has more arguments).

So, in Utility Functions we could instead add something like:

These functions are intended to be called in user-written code. Note that if a library requires a logging-function it is recommended to use a wrapper function calling e.g., ModelicaFormatMessage - allowing writing additional information, handling additional arguments, and compiling the library without access to a tool-specific ModelicaUtilities.h.

Reply:
OK.  I'll make a separate PR for recommended ways of using _ModelicaUtilities.h_.

Reply:
OK now?